### PR TITLE
Add type hints for libcoap.py

### PR DESCRIPTION
**Pull request by ggravlingen**

### CODE REVIEW

File: pytradfri/api/libcoap_api.py [hunk 1]
```diff
@@ -206,22 +221,5 @@ def _process_output(output, parse_json=True):
     if output.startswith(SERVER_ERROR_PREFIX):
         raise ServerError(output)
     if not parse_json:
-        return output
-    return json.loads(output)
-
-
-def retry_timeout(api, retries=3):
```
Comment (ggravlingen) [inline pytradfri/api/libcoap_api.py:213]: I can't see that this is being used anywhere? I thus removed it. Maybe do that in a separate PR and then look at adding type hints?

File: pytradfri/api/libcoap_api.py [hunk 2]
```diff
@@ -21,25 +22,34 @@
...
 
     @psk.setter
-    def psk(self, value):
+    def psk(self, value: str) -> None:
         """Set psk."""
         self._psk = value
 
     def _base_command(self, method: str) -> list[str]:
         """Return base command."""
+        if self._psk is None:
```
Comment (ggravlingen) [inline pytradfri/api/libcoap_api.py:50]: Not sure about this one, but `_psk` will not always be a string so we have to handle that somehow.

File: pytradfri/api/libcoap_api.py [hunk 3]
```diff
@@ -98,7 +110,9 @@ def _execute(self, api_command, *, timeout=None):
         api_command.process_result(_process_output(return_value, parse_json))
         return api_command.result
 
-    def request(self, api_commands, *, timeout=None):
+    def request(
+        self, api_commands: list[Command[T]], *, timeout: int | None = None
+    ) -> T | list[T] | list[Any] | None:
         """Make a request. Timeout is in seconds."""
         if not isinstance(api_commands, list):
```
Comment (ggravlingen) [inline pytradfri/api/libcoap_api.py:117]: Do we need this? Don't we expect api_commands to always be a list given the type hint for `api_commands`?

File: pytradfri/api/libcoap_api.py (hunk 2 above)
Comment (MartinHjelmare) [reply pytradfri/api/libcoap_api.py:50]: Raise a RuntimeError. The user needs to generate a key first.

File: pytradfri/api/libcoap_api.py (hunk 3 above)
Comment (MartinHjelmare) [reply pytradfri/api/libcoap_api.py:117]: Check the typing of the async api. It should be the same, including overrides.

File: pytradfri/api/libcoap_api.py (hunk 1 above)
Comment (MartinHjelmare) [reply pytradfri/api/libcoap_api.py:213]: I think we can remove it. I think it was used a long time ago as a helper.

File: pytradfri/api/libcoap_api.py (hunk 3 above)
Comment (ggravlingen) [reply pytradfri/api/libcoap_api.py:129]: I add to make some adjustments as `_execute` can also return `None`.

Comment (MartinHjelmare) [reply pytradfri/api/libcoap_api.py:129]: No, execute should always return a command result. When starting an observation we should also return the command result.

File: pytradfri/api/libcoap_api.py [hunk 4]
```diff
@@ -180,15 +206,18 @@ def generate_psk(self, security_key):
             self._psk = security_key
 
             # Ask the Gateway to generate the psk for the identity.
-            self._psk = self.request(Gateway().generate_psk(existing_psk_id))
+            command: list[Command[str]] = [Gateway().generate_psk(existing_psk_id)]
```
Comment (MartinHjelmare) [inline pytradfri/api/libcoap_api.py:209]: Revert this change.

**Review by MartinHjelmare — APPROVED** [review]
Great!
